### ISSUE TITLE
Failing to Compile with LaTeXMK

### ISSUE THREAD

**Issue by fuzzybear3965** [lervag/vimtex#235]
I understand that vimtex is configured to compile continuously. It doesn't do that. Issuing the command VimtexCompile or VimtexCompileSS or \ll all result in the following string of errors.

```
Error detected while processing function vimtex#latexmk#compile..vimtex#util#execute:
line   66:
E34: No previous command
```

No other errors are found when starting Vim. I am using the cream version of Vim on Windows. The following is the output of \li:

```
b:vimtex : test
  pid : 0
  out : ''
  tmp : 'C:\Users\JOHNRI~1\AppData\Local\Temp\VIB3D0E.tmp'
  root : 'C:\Users\John Rinehart\Documents'
  aux : 'C:\Users\John Rinehart\Documents/test.aux'
  log : 'C:\Users\John Rinehart\Documents/test.log'
  tex : 'C:\Users\John Rinehart\Documents\test.tex'
  cmd_latexmk_compile : 'cd /D "C:\Users\John Rinehart\Documents" && set max_print_line=2000 & latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -e "$success_cmd = ''\"vim\" --servername JOHN --remote-expr \"vimtex\#latexmk\#callback(1)\"''" -e "$failure_cmd = ''\"vim\" --servername JOHN --remote-expr \"vimtex\#latexmk\#callback(0)\"''" -e "$pdf_previewer = ''SumatraPDF ''" "test.tex"'
  base : 'test.tex'
  viewer
    init : function('6')
    view : function('7')
    latexmk_append_argument : function('8')
```

**Comment by lervag** [comment]
Interesting. I've never seen this particular problem before. Unfortunately, I don't work in windows (at all), and so I am not able to test, and I also don't know that much about what the problem might be.

What I do know, is that the commands are run in a windows command prompt (or whatever it is called). That means that you should be able to run most of the command seen in `cmd_latexmk_compile` in a command window. Thus, my suggestion is that you open a command window and try the different commands there.

Another thing you can try is to open the file `\path\to\vimtex\autoload\vimtex\util.vim` and go to line 101 (which is the same line as the error mentions). Here you can add a line, e.g.

```vimL
    if silent
      " Add line to echo the cmd before trying to execute
      echom "My command: " . cmd
      silent execute '!' . cmd
    else
      execute '!' . cmd
    endif
```

Now when you try to compile the command should be echoed. It might dissapear to quickly to see, in which case you can see the message again by using the command `:messages`.

**Comment by fuzzybear3965** [comment]
Thanks, Karl, for getting back to me so promptly. I've inserted the ``echom`` line into line 101 of ``util.vim`` and I used ``:messages`` to view the output. The following command was spat back at me:

    start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents" && set max_print_line=2000 
    & latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -pvc -e
    "$success_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(1)\"'" -e
    "$failure_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(0)\"'" -e
    "$pdf_previewer = 'SumatraPDF '" "test.tex"
    >C:\Users\JOHNRI~1\AppData\Local\Temp\VIA145B.tmp""

**Comment by lervag** [comment]
First: The servername comes from the internal variable `v:servername`, and it is set automatically by Vim (defaults to VIM or GVIM).

I am happy to hear that the command works as expected on the command line. This implies that things should work.

So, first things first:

1. Which OS are you on? Windows, but which version.
2. What are your vimtex settings?

Finally, could you please test with a minimal vimrc file? That is, something like the following should work:

```vimL
set nocompatible
let &rtp = '~/.vim/bundle/vimtex,' . &rtp
filetype plugin indent on
syntax enable

"
" vimtex options here
"
```

Note that the path to vimtex must be right. The path I chose is a typical install path, but it might be different on your end (particularly since your on Windows).

**Comment by fuzzybear3965** [comment]
I definitely think it will work soon.

   1.  I am on Windows 8.1 64-bit.
   2.  I think when you ask for my vimtex settings you are asking about what things I have specified in my vimrc file. To make this clear, then, I will just include my vimrc file below:

    set nocompatible
    let &rtp = '~/.vim/bundle/vimtex,' . &rtp
    filetype plugin indent on
    syntax enable

    let &rtp = 'C:\cream\vim74,' . &rtp

    " Set the runtime path to include Vundle and initialize
    set rtp+=~/.vim/bundle/Vundle.vim
    call vundle#begin()
    " let Vundle manage Vundle, required
    Plugin 'gmarik/Vundle.vim'
    Plugin 'lervag/vimtex'
    call vundle#end()

    " vimtex settings
    let g:vimtex_view_general_viewer = 'SumatraPDF'

For whatever reason I couldn't get Markdown to work in the above. I'm sorry about that. Additionally, I'll include my vimtex path and my runtime path below.

Vimtex path: C:\Users\John Rinehart\.vim\bundle\vimtex (inside of this directory autoload, ftplugin, etc. is included)
&rtp: C:\Users\John Rinehart/vimfiles,C:\Users\John Rinehart\.vim\bundle\Vundle.vim,C:\Users\John Rinehart\.vim\bundle\vimtex,C:\cream\vim74,C:\Users\John Rinehart\.vim\bundle\vimtex,C:\cream/vimfiles,C:\cream\vim74,C:\cream/vimfiles/after,C:\Users\John Rinehart/vimfiles/after,C:\Users\John Rinehart/.vim/bundle/Vundle.vim,C:\Users\John Rinehart\.vim\bundle\Vundle.vim/after,C:\Users\John Rinehart\.vim\bundle\vimtex/after

**Comment by fuzzybear3965** [comment]
Okay, I have made some progress! I was wrong earlier when I said that the command worked.  The command compiles the document and renders a .PDF but it also returns an error that I didn't see because I got too excited. The following is the command that I see in Vim when I try to compile a document. I got this command from using your line, Karl:
```
start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents" && set max_print_line=2000 &
 latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -pvc -e 
"$success_cmd = '\"vim\" --servername VIM1 --remote-expr \"vimtex\#latexmk\#callback(1)\"'" -e 
"$failure_cmd = '\"vim\" --servername VIM1 --remote-expr \"vimtex\#latexmk\#callback(0)\"'" -e 
"$pdf_previewer = 'SumatraPDF '" "test.tex" 
>C:\Users\JOHNRI~1\AppData\Local\Temp\VIA18F6.tmp""
```
Running this command in the Windows command prompt yields the following output when I'm outside of the directory that contains the file that I'm trying to compile:
```
Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.

------------
Latexmk: Could not find file [test.tex].
-- Use the -f option to force complete processing.
```
When I navigate to the directory that contains the file and I execute the same command:
```
 Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.
Viewing pdf
ps: user John unknown
Latexmk: I have not found a previewer that is already running.
   So I will start it for 'test.pdf'
------------
------------
Running 'start SumatraPDF   "test.pdf"'
------------
------------
Running 'vim --servername VIM --remote-expr "vimtex\#latexmk\#callback(1)"'
------------
E247: no registered server named "VIM": Send expression failed.
```
This is pretty easy to fix. I imagine that this error is only occurring because I'm running this on the command line and no instance of Vim is currently running. So, if I run the following command in a second command line window ``vim --servername vim`` and run the same command as originally stated in the first command window I receive the following error:
```
Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.
Viewing pdf
ps: user John unknown
Latexmk: I have not found a previewer that is already running.
   So I will start it for 'test.pdf'
------------
------------
Running 'start SumatraPDF   "test.pdf"'
------------
------------
Running 'vim --servername VIM --remote-expr vimtex#latexmk#callback(1)'
------------
sh: -c: line 0: syntax error near unexpected token `('
sh: -c: line 0: `vim --servername VIM --remote-expr vimtex#latexmk#callback(1)'
```
At this point it looks like an escaping issue. So, I've played around with it a bit and I don't get any errors when I execute the following command:
```
start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents" && set max_print_line=2000 & 
latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -pvc -e 
"$success_cmd = '\"vim\" --servername VIM --remote-expr \"\"vimtex#latexmk#callback(1)\"\"'" -e 
"$failure_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(0)\"'" -e 
"$pdf_previewer = 'SumatraPDF '" "test.tex" 
>C:\Users\JOHNRI~1\AppData\Local\Temp\VIA281F.tmp""
```
The difference being that in the "success_cmd" section of the latexmk command being executed I have added a set of escaped quotes around ``vimtex#latexmk#callback(1)`` and I have removed the escaping backslash from the ``#``s. This results in the following in the command window:
```
Latexmk: This is Latexmk, John Collins, 5 February 2015, version: 4.43a.
Viewing pdf
ps: user John unknown
Latexmk: I have not found a previewer that is already running.
   So I will start it for 'test.pdf'
------------
------------
Running 'start SumatraPDF   "test.pdf"'
------------
------------
Running 'vim --servername VIM --remote-expr "vimtex#latexmk#callback(1)"'
------------
```
The state of the command window doesn't change at this point. It just "hangs". I think the program is not exiting because latexmk is waiting for changes in the file and your callback is keeping the program alive. But, I honestly don't know why my cursor just sits below that last line and blinks.

**Comment by lervag** [comment]
Ok, now there are a lot of things going on. I'll begin with the vimrc file: The vimrc file you supplied is somewhat confused. In particular, it combines manually changing `&rtp` with the use of a plugin manager (Vundle). I propose that you simplify further and instead just use this:

```vimL
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable
let g:vimtex_view_general_viewer = 'SumatraPDF'
```

If things now work as expected, then we know the problem is due to a plugin conflict. Else there is either a bug or some other problem that needs fixing on the vimtex side.

----

Now continue to the suggestion about the addition of more escapes. If I was able to parse correctly, you suggest the following change to the command to get properly escaping:

```
# Old    \"vimtex\#latexmk\#callback(1)\"'"
# New    \"\"vimtex#latexmk#callback(1)\"\"'"
```

However, I believe the command should already be properly escaped. It has been tested and verified by several other windows users, such as @Konfekt, @nasenamer and @LeszekSwirski.

As you probably understand already, there is quite alot of things that might go wrong. I therefore suggest that we simplify by disabling the callback feature. Please add the following option to your vimrc file and test if things now work:

```vimL
let g:vimtex_latexmk_callback = 0
```

That is, test whether the `<leader>ll` key works as expected. Note that with the callback disabled, there will be no message on successful/erroneous compilation. The command should only give a message that compilation has started in continuous mode.

----

Based on one of the final sentences in your last comment, it seems you have not completely understood the `latexmk -pvc` command, which is what is used by vimtex to start continuous compilation. This command starts latexmk in such a way that it will run "forever" (until it is stopped/killed). Whenever one of the source files is updated, it will automatically compile the document again. Thus, the command window is supposed to "hang".

**Comment by fuzzybear3965** [comment]
Okay, let's submit my daily "Karl" report. I really didn't think that the Vundle plugin could be interfering but I've removed those lines from my _vimrc file. It now reads as follows:
```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

let g:vimtex_view_general_viewer = 'SumatraPDF'
```
Trying to compile this generates the following messages:
```
My command: start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents"
 && set max_print_line=2000 
& latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" 
-pvc -e "$success_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(1)\"'" 
-e "$failure_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(0)\"'" 
-e "$pdf_previewer = 'SumatraPDF '" "test.tex" >C:\Users\JOHNRI~1\AppData\Local\Temp\VIABF56.tmp""

Error detected while processing function 
vimtex#latexmk#toggle..vimtex#latexmk#compile..vimtex#util#execute:
line   68:
E34: No previous command
```
We have seen these messages before. 

----
You parsed those two commands perfectly. I should have summarized it for you like you did for me. I really think it's an escaping issue but, you're right, I would expect to see this with all Windows users.

I have disabled the callback feature such that I, now, have the following ``_vimrc`` file: 
```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

let g:vimtex_view_general_viewer = 'SumatraPDF'

let g:vimtex_latexmk_callback = 0
```
Sourcing my ``_vimrc`` file and editing my TeX file (``test.tex``) now results in a similar, but not identical, message as before. Note that, now, as expected, there are no remote callback expressions being sent to a vim server.
```
My command: start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents" 
&& set max_print_line=2000 
& latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -pvc 
-e "$pdf_previewer = 'SumatraPDF '" "test.tex" >C:\Users\JOHNRI~1\AppData\Local\Temp\VIA38FC.tmp""
Error detected while processing function 
vimtex#latexmk#toggle..vimtex#latexmk#compile..vimtex#util#execute:
line   68:
E34: No previous command
```
Note that no matter if the global variable ``g:vimtex_latexmk_callback`` is set to 0 inside my ``_vimrc`` file or not the .PDF doesn't compile and the following message is spat out in the base of my vim instance: ``latexmk compile: started continuous mode``. So, it does what you thought it would. The only thing that changes is the inclusion/exclusion of the remote callback expressions in the ``latexmk`` perl commands.

---
What you said regarding the ``-pvc`` file makes sense.

**Comment by fuzzybear3965** [comment]
I have looked at issue #221 and I my _vimrc file now looks like:
```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

"let &rtp = 'C:\cream\vim74,' . &rtp

let g:vimtex_view_general_viewer = 'SumatraPDF'
let g:vimtex_view_general_options ='-forward-search @tex @line @pdf'
let g:vimtex_view_general_options_latexmk='reuse-instance'
let g:vimtex_latexmk_continuous = 0
let g:vimtex_latexmk_callback = 0
```
Trying to ``\ll`` now generates the following message at the base of Vim: ``latexmk compile: compiling ...``. That message does not go away or terminate (change state). In addition to this text at the bottom of Vim I also have the following messages seen when I execute ``:messages``:
```
Error detected while processing function vimtex#latexmk#toggle..vimtex#latexmk#compile..vimtex#util#execute:
line   70:
E34: No previous command
```

**Comment by jbkopecky** [comment]
I am not sure if it is my config or not, but I have the same issue since commit https://github.com/lervag/vimtex/commit/8e3cefbe940a3c7cc33386d30518f5ad67210531, I have messed with it a little and I got able to get vimtex to compile again, but I am not sure what happened.

**Comment by lervag** [comment]
@jbkopecky This is not really helpful: I have no idea what your config is, I do not know what you mean by "Messed with it". I don't even know what your problem is. Do you have the same problem as @fuzzybear3965? I don't mean to be rude, but it would be much more helpful if you could be more precise and informative: Please explain what the problem is (or refer to the original issue if you have the same problem), and please provide the configuration. Also, please test with a minimal vimrc file, e.g. something similar to what @fuzzybear3965 presents in the latest comment.

@fuzzybear3965 The message will go away when a different message appears or when you do a rewrite (this is how messages work in Vim). Could you repeat once more with the same vimrc file. Do the following:

1. Open your test file
2. Do `\lC` to remove all generated files (this does a `latexmk -C`)
3. Do `\ll` (report output)
4. Report output of `:messages`

Also, what is the content of your `.latexmkrc` file? I do not understand why you get `dvi` files. Vimtex should compile to `pdf`.

**Comment by jbkopecky** [comment]
Hello, sorry to be not so informative: I have had exactly the same issue as @fuzzybear3965 for a few months now on my windows machine. I will try to find time to be more specific asap.

**Comment by lervag** [comment]
Thanks! I am curious, it seems you changed the commit reference. I guess it might have been a typo. But in any case, did things work as expected before this commit, and so is this the first commit that breaks things for you?

**Comment by jbkopecky** [comment]
Yes, sorry, it was a typo. It did work perfectly before. I am not entirely sure it is exactly this commit that broke it for me, but I found it searching in vimtex code line by line for the issue that sends this error message. I will make sure to try to understand what exactly happens here and try to post a fix that works for me.

**Comment by fuzzybear3965** [comment]
@lervag and @jbkopecky, I'm glad there's some renewed interest in this! I've been waiting to try anything so that I wouldn't modify my vimtex state, too much. @jbkopecky figure out what you did :)!

Okay, @lervag, you have said to use the "same vimrc file". Based on issue #221 I have two different ``_vimrc`` files that I have configured. Removing the bottom 3 options (latexmk general options, continuous and callback) and doing the same test you suggested produce different results. The first set of results follows and uses the following ``_vimrc`` file.

#### _vimrc 1
---

```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

"let &rtp = 'C:\cream\vim74,' . &rtp

let g:vimtex_view_general_viewer = 'SumatraPDF'
let g:vimtex_view_general_options_latexmk='reuse-instance'
let g:vimtex_latexmk_continuous = 0
let g:vimtex_latexmk_callback = 0
```

I've performed the ``\lC`` (vim status: ``latexmk clean: finished (full)``). Running `:messages` after this yields: 

```My command: start /b "cd /D "C:\Users\John Rinehart\Documents" & latexmk -C "test.tex" >nul"
shell returned 1```

 I've also performed a `\ll`. The output is different now. Vim doesn't error out. Instead, the command prompt (read: shell) reports the following

```
Latexmk: Stopping because executing following code from command line
    $pdflatex =~ s/^((.(?<start /b cd
gave an error:
    Substitution replacement not terminated at (eval 8) line 1.

The filename, directory name, or volume label syntax is incorrect.

shell returned 1

Press ENTER or type command to continue
```
Pressing ``ENTER`` brings me back to Vim where ``:messages`` reports no additional messages.

#### _vimrc 2
---
```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

"let &rtp = 'C:\cream\vim74,' . &rtp

let g:vimtex_view_general_viewer = 'SumatraPDF'
```
Running ``lC`` works. Now, running ``\ll`` changes the Vim status to ``latexmk compile: started continuous mode``. Running ``:messages`` after executing both commands yields:

```
My command: start /b "cd /D "C:\Users\John Rinehart\Documents" & latexmk -C "test.tex" >nul"
My command: start /b "cmd /s /c "cd /D "C:\Users\John Rinehart\Documents" && set max_print_line=2000 & latexmk -pdf -e "$pdflatex =~ s/^((.(?<!^internal(?=\\s)))*?) /$1 -file-line-error /" -pvc -e "$success_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(1)\"'" -e "$failure_cmd = '\"vim\" --servername VIM --remote-expr \"vimtex\#latexmk\#callback(0)\"'" -e "$pdf_previewer = 'SumatraPDF '" "test.tex" >C:\Users\JOHNRI~1\AppData\Local\Temp\VIA89F0.tmp""
```
---
With regards to your question concerning a ``.latexmkrc`` file. I do not have a ``.latexmkrc`` or ``_latexmkrc`` file. But, I have ran ``latexmk`` from the command line as follows: ``latexmk test.tex`` and it generates DVI files by default (no PDFs). Note that when I run ``\lC`` from the command line it successfully clears the generated files from the working directory. I have triet ``lC`` after executing ``latexmk`` from the command line manually (it compiles manually from command line to DVI, generating other compile-related files too).

Also, note my correction to my previous post from 4 days ago: 

```I have checked and there are **not** compile-related files (test.aux,test.log,test.dvi, etc.) in the working directory. So, it doesn't seem to be actually compiling.```

Thank you all.

**Comment by lervag** [comment]
@fuzzybear3965 Thank you! The lines with `My command` are not from vimtex directly, they come after the earlier comment ([see here](https://github.com/lervag/vimtex/issues/235#issuecomment-144536651)).

Could we continue with *_vimrc 2* for the moment? I know you prefer to not use continuous mode, but I would be happy if we could solve the problem with the default mode first. I also suspect that the problem is the same in both cases, and so if we fix the problem with *_vimrc 2*, then it should work with *_vimrc 1* (famous last words).

@fuzzybear3965 and @jbkopecky Could either of you please test whether things work correctly if you revert the change that was brought in 41f0eccdc0b9f3cb80e5246dfe07f646d3aa0768? The diff should make it simple to see the required changes.

@jbkopecky Do you know about git bisect? It is a very useful tool that we can use to pinpoint the exact changeset that breaks things. See [here](https://git-scm.com/docs/git-bisect) for the man page. I would be happy if you would be able to use this (or of course simple manual testing) to verify if it is indeed 41f0eccdc0b that breaks things.

**Comment by lervag** [comment]
Ok, I've installed vim on Windows, together with MikTeX, SumatraPDF and perl. I've done a git bisect, and I've found that the problem seems to come from 95e30c706, and it is related to #212.

**Comment by lervag** [comment]
I think I've found the problem. Please test latest version and see if things work. It should work with both of the vimrc files.

**Comment by fuzzybear3965** [comment]
@lervag, you're the man! (I had understood that the ``My command`` outputs were due to the debug statement I added. But, don't they come from vimtex in that this is the program that generates these shell/command prompt commands?)

It's not true that I prefer not to use continuous mode. I'm down for continuous mode! I have (you have) fixed the problem with *_vimrc 2* by rolling back to commit ``47ac958``. I tried the next few commits after this one. None of those (and the commit that I was on, previously) don't work.  So, I'll play around with vimtex. But, it seems that at least on this commit, I'm okay. I looked at the diff of the two commits and I see the addition of the ``internal`` in the regex. Why would this be affecting me? What's happening?

However, I want to note that once I change my ``_vimrc`` file to include the option ``let g:vimtex_latexmk_continuous = 0``, I receive the following error from the command line (not Vim)
```

\"'" was unexpected at this time.

shell returned 1

Press ENTER or type command to continue
```
I tell you this more for completeness than for anything. I won't be using this option (continuous all the way!). The full ``_vimrc`` for this condition is as follows:

```
set nocompatible
let &rtp = 'C:\Users\John Rinehart\.vim\bundle\vimtex,' . &rtp
filetype plugin indent on
syntax enable

"let &rtp = 'C:\cream\vim74,' . &rtp

let g:vimtex_view_general_viewer = 'SumatraPDF'
let g:vimtex_latexmk_continuous = 0
```
